### PR TITLE
PHP: Make the default package more sane [v3]

### PR DESCRIPTION
**Pull request by etu**
###### Motivation for this change
Our old PHP derivation was crazy big with a crazy amount of enabled extensions that also had a crazy amount of defaults that we pulled in, yet excluded some extensions that would have been nice to have by default.

This is a follow up on: #82348 to utilize #79377 more. And also a follow-up on all of the work done by me and @talyz in #82794 and #83514

But this grew a bit more, and was more complex than I imagined when I started this. But it became a collaboration project with @talyz to finish everything up.

We have also written the initial documentation for PHP in NixOS.

#### The current PHP
The current closure size is:
```
/nix/store/4ga497cqvz04vcyw9lf7s74ymrc6r28g-php-7.4.3	234.3M
````

And this contains things (among others) like:
```
/nix/store/y5166rf2lkxkf3991qm390vn92czgfgh-openldap-2.4.49          	 53.6M
/nix/store/2zsgfrkp0hn878rnrp0azpmy76ljm2b0-linux-pam-1.3.1          	 53.6M
```

#### The new smaller PHP
Here we have a smaller closure size:
```
/nix/store/10867i6grp8w8p6p6k1w8rxiwza5nyyy-php-with-extensions-7.4.3	189.4M
```

And we have a smaller base package:
```
/nix/store/v5v6wn207ibpp6n90z05f11k7svq94wd-php-7.4.3	 85.5M
```

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by etu** [comment]
@GrahamcOfBorg build php72 php73 php74 php72Packages.exts php73Packages.exts php74Packages.exts nixosTests.nextcloud nixosTests.php

**Review by aanderse — COMMENTED** [review]
I haven't been through all of this yet, but here are a few initial suggestions for wording in the documentation if anyone find them useful.

File: pkgs/top-level/php-packages.nix
Comment (Izorkin) [inline pkgs/top-level/php-packages.nix:624]: Pthreads and pinba phpPackages removed?

File: doc/languages-frameworks/php.section.md
Comment (jtojnar) [inline doc/languages-frameworks/php.section.md:20]: `php.phpPackages` feels too verbose. Could we just do `php.packages` or `php.pkgs` like Python does?

File: pkgs/top-level/php-packages.nix
Comment (etu) [reply pkgs/top-level/php-packages.nix:624]: For now yes because they did not build. Whenever this change is pulled in and build it's easier to add things back and actually test it in a better way.

We reasoned that we had a set of changes that was big enough as it was.

So for now I think that's best, but I won't reject adding it back in a better way if they build.

File: doc/languages-frameworks/php.section.md
Comment (jtojnar) [inline doc/languages-frameworks/php.section.md:59]: Would it make sense to mirror other language infrastructures like Python, which use `withPackages`?

There is also https://discourse.nixos.org/t/wip-rfc-a-new-nixpkgs-frontend-for-language-infrastructures/3447

Comment (etu) [reply doc/languages-frameworks/php.section.md:59]: It would, `php.buildEnv` does more things than just adding packages though. And it's something that already exists in `master`. So just renaming it wouldn't make sense. But we can probably add a wrapper for `buildEnv` named `withPackages`. But it's not a super big thing and could be done later on as well.

File: doc/languages-frameworks/php.section.md
Comment (talyz) [reply doc/languages-frameworks/php.section.md:20]: Good idea! I'll fix that right away.

File: doc/languages-frameworks/php.section.md
Comment (talyz) [reply doc/languages-frameworks/php.section.md]: I've addressed all of these and added some additional fixes, so please have another look :)

File: doc/languages-frameworks/php.section.md
Comment (talyz) [reply doc/languages-frameworks/php.section.md]: Done!

**Comment by talyz** [comment]
@GrahamcOfBorg build php72 php73 php74 php72Packages.exts php73Packages.exts php74Packages.exts nixosTests.nextcloud nixosTests.php

File: doc/languages-frameworks/php.section.md
Comment (aanderse) [reply doc/languages-frameworks/php.section.md:62]: I thought `withPackages` was explicitly rejected by someone... (@FRidh maybe) because that would be misleading. I'm having trouble revealing the details at the moment.

Comment (FRidh) [reply doc/languages-frameworks/php.section.md:62]: My remark was that `withPackages` should only take a lambda, not an attribute set
https://github.com/NixOS/nixpkgs/pull/79377#issuecomment-583104499

Comment (aanderse) [reply doc/languages-frameworks/php.section.md:62]: @FRidh ah, thanks for clarifying!

Comment (etu) [reply doc/languages-frameworks/php.section.md:62]: I think a `withPackages` wrapper can be added later on after this PR as well. It's a quite small thing to add to make it simpler to use. But for now I'm fairly happy with all the progress we have here.

We have a bunch of small things we would like to improve after this PR. But we had enough changes already and tried to stick to "a minimal" amount of changes.

We (me and @talyz) were discussing how we possibly could break it down into several PRs and came to the conclusion that we can't.

**Comment by aanderse** [comment]
My only remaining comment is that I wonder if it becomes confusing (or overly verbose) to have `php.packages.exts`, especially for people newer to nixpkgs. Currently in my mind I wonder if `php.packages` and `php.extensions` (or `php.pkgs` and `php.exts`) is less confusing, but of course that would potentially end up making `buildEnv` more verbose. To alleviate that maybe we could have `buildEnv` operate on `php` instead of `pp`, like `php.buildEnv { exts = with p; [ pkgs.imagick exts.opcache ]; }`... but maybe I'm just rambling at this point and this doesn't make any sense... feel free to disregard :smile:

**Comment by talyz** [comment]
@aanderse Yes, that's also something we discussed, but decided that the scope of the PR was big enough as is. Definitely something to think about, though :)

File: pkgs/top-level/all-packages.nix
Comment (flokli) [inline pkgs/top-level/all-packages.nix:9390]: I very much approve inlining all these into the only derivations that use it, but we still need to add the attributes to `pkgs/top-level/aliases.nix` with a throw explaining they were inlined, and how to obtain a similarily-configured PHP.

**Review by flokli — COMMENTED** [review]
Only read over the docs and skimmed over the code so far, haven't actually run anything yet.

**Comment by FRidh** [comment]
> Currently in my mind I wonder if php.packages and php.extensions (or php.pkgs and php.exts) is less confusing, but of course that would potentially end up making buildEnv more verbose.

I think we should typically go for the verbose options. Readability is important. Now, we already use `pkgs` everywhere, so maybe not change that one, but `exts` is...strange.

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (etu) [reply nixos/doc/manual/release-notes/rl-2009.xml:130]: Fixed :)

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (etu) [reply nixos/doc/manual/release-notes/rl-2009.xml]: Fixed :-)

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (etu) [reply nixos/doc/manual/release-notes/rl-2009.xml]: I've clarified on this now :-)

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (etu) [reply nixos/doc/manual/release-notes/rl-2009.xml]: Fixed as well :-)

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (etu) [reply nixos/doc/manual/release-notes/rl-2009.xml]: I've re-written this to make more sense.

The project to do this has been big we've gone back and forth on a bunch of things a couple of times while working on it. It has resulted in things (especially descriptions) that may be outdated here and there. :-)

**Comment by etu** [comment]
> > Currently in my mind I wonder if php.packages and php.extensions (or php.pkgs and php.exts) is less confusing, but of course that would potentially end up making buildEnv more verbose.
> 
> I think we should typically go for the verbose options. Readability is important. Now, we already use `pkgs` everywhere, so maybe not change that one, but `exts` is...strange.

Yeah, but since there's a lot of extensions from upstream and a smaller set of extensions from 3rd party sources. We kinda wanted to make a distinction between the two types since that can indicate how reliable and maintained they are.

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (etu) [reply nixos/doc/manual/release-notes/rl-2009.xml:157]: They are needed, I've just tried to remove the `para` block and then the manual doesn't build anymore :-)

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:22]: Is `opcache.enable = 1` not needed anymore? This isn't part of `extraConfig` anymore AFAICS.

**Comment by talyz** [comment]
> > Currently in my mind I wonder if php.packages and php.extensions (or php.pkgs and php.exts) is less confusing, but of course that would potentially end up making buildEnv more verbose.
> 
> I think we should typically go for the verbose options. Readability is important. Now, we already use `pkgs` everywhere, so maybe not change that one, but `exts` is...strange.

I agree - this should be changed. Since we now have `php.packages` I don't see why we wouldn't move `php.packages.exts` to `php.extensions`. However, just to be clear: `exts` was not introduced in this PR - it only moves `phpPackages`, which `exts` was already a part of, to `php.packages`.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:18]: Have you confirmed that this config is actually taken into account by phpfpm entirely? The module adds another `-c php.ini` in the systemd-unit which IIRC overrides changes from `php.buildEnv`:

https://github.com/NixOS/nixpkgs/blob/24d456a48daccf1f8b179c53d92ed3c2c3bc47b4/nixos/modules/services/web-servers/phpfpm/default.nix#L276

File: pkgs/top-level/php-packages.nix
Comment (Ma27) [inline pkgs/top-level/php-packages.nix:483]: Are we still using this approach somewhere to ensure that an extension is only built against supported PHP versions? If yes, we can probably do some sort of `disabled = <cond>` (as it's done in python packages) IMHO.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (talyz) [reply nixos/modules/services/web-apps/nextcloud.nix:22]: Opcache is enabled by default if the module is loaded, according to the [PHP website](https://www.php.net/manual/en/opcache.configuration.php). I'll check if this is the case when running the binary.

Comment (talyz) [reply nixos/modules/services/web-apps/nextcloud.nix:22]: It is indeed on by default.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (talyz) [reply nixos/modules/services/web-apps/nextcloud.nix:18]: That's why `php.buildEnv` was rewritten to not use the `-c`-flag. It instead uses an environment variable to set additional config.

File: pkgs/top-level/php-packages.nix
Comment (talyz) [reply pkgs/top-level/php-packages.nix:483]: I got rid of all uses in `php-packages.nix` by replacing them with `meta.broken` or, in the case of `pinba` and `pthreads`, removing them. A `disabled` param would be nice, though :)

File: pkgs/top-level/all-packages.nix
Comment (etu) [reply pkgs/top-level/all-packages.nix:9390]: I've now made two fixup commits to add these to the aliases files, the commit ids are 9e51ba64d9e and 98b77d59ba0.

Me and @talyz can squash them in later :-)

**Comment by etu** [comment]
> > > Currently in my mind I wonder if php.packages and php.extensions (or php.pkgs and php.exts) is less confusing, but of course that would potentially end up making buildEnv more verbose.
> > 
> > 
> > I think we should typically go for the verbose options. Readability is important. Now, we already use `pkgs` everywhere, so maybe not change that one, but `exts` is...strange.
> 
> I agree - this should be changed. Since we now have `php.packages` I don't see why we wouldn't move `php.packages.exts` to `php.extensions`. However, just to be clear: `exts` was not introduced in this PR - it only moves `phpPackages`, which `exts` was already a part of, to `php.packages`.

It would be nice to separate extensions from command line utilities... It would make it much nicer :D

**Comment by etu** [comment]
@GrahamcOfBorg eval

**Comment by etu** [comment]
@GrahamcOfBorg build php72 php73 php74 php72.extensions php73.extensions php74.extensions php72.packages php73.packages php74.packages nixosTests.nextcloud nixosTests.php unit uwsgi

**Comment by etu** [comment]
@GrahamcOfBorg eval

**Comment by talyz** [comment]
@jtojnar @aanderse I've now added a `php.withExtensions` to provide a simpler interface when the additional config option provided by `php.buildEnv` is unnecessary.

I also renamed exts to extensions in `php.buildEnv` and rewrote some of the documentation.

Me and @etu now feel like this is done and that all raised concerns have been addressed.

**Review by Ma27 — APPROVED** [review]
Awesome work :+1:

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (flokli) [inline nixos/modules/services/web-apps/nextcloud.nix:15]: I'd assume these are the default extensions, plus the additional `acpu`, `redis`, `memcached` and `imagick` extensions - or did you go through the requirement list of nextcloud?

If it's just the default extensions, plus some additional ones, could we expose the default ones, maybe via `pkgs.php74.defaultExtensions`? That'd make things way more obvious.

I know it's less explicit, but should

**Review by flokli — APPROVED** [review]
LGTM! One minor UX question, but no reason to block this PR.

**Review by aanderse — APPROVED** [review]
I played around with this today and it worked well :+1:

@jtojnar are you good with this?
@etu @talyz thanks for being superstars!

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (etu) [reply nixos/modules/services/web-apps/nextcloud.nix:15]: You're exactly right. It's the default list plus apcu, redis, memcached and imagick.

After we get this in I could make that available and document how to extend PHP based on the default set. I think our changeset is big enough as it is :-)

Comment (etu) [reply nixos/modules/services/web-apps/nextcloud.nix:15]: @flokli How would you like a `php.withAdditionalExtensions` or `php.withExtraExtensions` that takes the default list and adds extra ones? :)

Of course we could expose the default list as well.

Comment (flokli) [reply nixos/modules/services/web-apps/nextcloud.nix:15]: Disclaimer: This disucssion should not block the PR. It's good to go IMHO!

Exposing the list of default extensions is probably already by itself quite readable and enough. A `withAdditionalExtensions` doesn't add too much benefit IMHO, and wouldn't allow me to do things like: take all the default extensions, but remove module x, and replace it with a patched version of it…

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (jtojnar) [inline nixos/doc/manual/release-notes/rl-2009.xml:137]:
```suggestion
(pkgs.php.buildEnv { extenssions = pp: with pp; [
```

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (jtojnar) [inline nixos/modules/services/web-apps/nextcloud.nix:10]:
```suggestion
    extensions = pp: with pp; [
```

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (etu) [reply nixos/doc/manual/release-notes/rl-2009.xml]: Thanks! Fixed both! :)

**Review by jtojnar — APPROVED** [review]

**Comment by etu** [comment]
Since we have approval from several people that care about this ecosystem and addressed the concerns by people that care about similar ecosystems within nixpkgs I think we can call this good to merge as it is.

So if nobody speaks up before 20:00 (CEST) today (that is in 8:25 from the time of this posting) unless someone speaks up before then or just merge it themself before that :-)

File: pkgs/top-level/php-packages.nix
Comment (Izorkin) [reply pkgs/top-level/php-packages.nix:624]: This pakages builded with this patch

```
diff --git a/pkgs/top-level/php-packages.nix b/pkgs/top-level/php-packages.nix
index 21207c72fe9..7dd422e839d 100644
--- a/pkgs/top-level/php-packages.nix
+++ b/pkgs/top-level/php-packages.nix
@@ -18,6 +18,7 @@ let
     pname = "php-${pname}";
   });

+  isPhp72 = pkgs.lib.versionAtLeast php.version "7.2";
   isPhp73 = pkgs.lib.versionAtLeast php.version "7.3";
   isPhp74 = pkgs.lib.versionAtLeast php.version "7.4";

@@ -295,7 +296,7 @@ in
   # This is a set of PHP extensions meant to be used in php.buildEnv
   # or php.withExtensions to extend the functionality of the PHP
   # interpreter.
-  extensions = {
+  extensions = rec {
     apcu = buildPecl {
       version = "5.1.18";
       pname = "apcu";
@@ -578,6 +579,53 @@ in
       configureFlags = [ "--with-excel" "--with-libxl-incdir=${pkgs.libxl}/include_c" "--with-libxl-libdir=${pkgs.libxl}/lib" ];
     };

+    pinba = if isPhp73 then pinba112 else pinba111;
+
+    pinba111 = buildPecl {
+      version = "1.1.1";
+      pname = "pinba";
+
+      src = pkgs.fetchFromGitHub {
+        owner = "tony2001";
+        repo = "pinba_extension";
+        rev = "RELEASE_1_1_1";
+        sha256 = "1kdp7vav0y315695vhm3xifgsh6h6y6pny70xw3iai461n58khj5";
+      };
+
+      meta = with pkgs.lib; {
+        description = "PHP extension for Pinba";
+        longDescription = ''
+          Pinba is a MySQL storage engine that acts as a realtime monitoring and
+          statistics server for PHP using MySQL as a read-only interface.
+        '';
+        homepage = "http://pinba.org/";
+        broken = (isPhp73);
+      };
+    };
+
+    pinba112 = buildPecl {
+      version = "1.1.2-dev";
+      pname = "pinba";
+
+      src = pkgs.fetchFromGitHub {
+        owner = "tony2001";
+        repo = "pinba_extension";
+        rev = "edbc313f1b4fb8407bf7d5acf63fbb0359c7fb2e";
+        sha256 = "02sljqm6griw8ccqavl23f7w1hp2zflcv24lpf00k6pyrn9cwx80";
+      };
+
+      meta = with pkgs.lib; {
+        description = "PHP extension for Pinba";
+        longDescription = ''
+          Pinba is a MySQL storage engine that acts as a realtime monitoring and
+          statistics server for PHP using MySQL as a read-only interface.
+        '';
+        homepage = "http://pinba.org/";
+
+        broken = (!isPhp73);
+      };
+    };
+
     protobuf = buildPecl {
       version = "3.11.2";
       pname = "protobuf";
@@ -595,6 +643,40 @@ in
       };
     };

+    pthreads = if isPhp73 then pthreads32-dev else pthreads32;
+
+    pthreads32 = buildPecl rec {
+      version = "3.2.0";
+      pname = "pthreads";
+
+      src = pkgs.fetchFromGitHub {
+        owner = "krakjoe";
+        repo = "pthreads";
+        rev = "v${version}";
+        sha256 = "17hypm75d4w7lvz96jb7s0s87018yzmmap0l125d5fd7abnhzfvv";
+      };
+
+      buildInputs = with pkgs; [ pcre.dev ];
+
+      meta.broken = (!isPhp72 || isPhp73);
+    };
+
+    pthreads32-dev = buildPecl {
+      version = "3.2.0-dev";
+      pname = "pthreads";
+
+      src = pkgs.fetchFromGitHub {
+        owner = "krakjoe";
+        repo = "pthreads";
+        rev = "4d1c2483ceb459ea4284db4eb06646d5715e7154";
+        sha256 = "07kdxypy0bgggrfav2h1ccbv67lllbvpa3s3zsaqci0gq4fyi830";
+      };
+
+      buildInputs = with pkgs; [ pcre2.dev ];
+
+      meta.broken = (!isPhp73 || isPhp74);
+    };
+
     redis = buildPecl {
       version = "5.1.1";
       pname = "redis";

```

Comment (etu) [reply pkgs/top-level/php-packages.nix:624]: Pinba is now added back in 1e42497447d385a87a3c8df21f269575d2061c3a.

A bit differently written but it's there. I will work on pthreads next.

Comment (etu) [reply pkgs/top-level/php-packages.nix:624]: Pthreads builds now for me on php72 and php73 but not php74. And it is archived by upstream: https://github.com/krakjoe/pthreads

So it doesn't seem to be supported or updated anymore. So I'll drop that one for now unless you have a patch for it that works on all versions.

Comment (etu) [reply pkgs/top-level/php-packages.nix:624]: Ok, added it as dccccab1e24 and marked it as broken for php74. Now it's back to how it was before :)

Comment (Izorkin) [reply pkgs/top-level/php-packages.nix:624]: > Pthreads builds now for me on php72 and php73 but not php74. And it is archived by upstream: https://github.com/krakjoe/pthreads

On php 7.4 package pthreads replaced to https://github.com/krakjoe/parallel

Comment (flokli) [reply pkgs/top-level/php-packages.nix:624]: I propose to do this in a followup PR. No need to delay this PR any further :-)

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (talyz) [reply nixos/modules/services/web-apps/nextcloud.nix]: I've added an `enabledExtensions` attribute and documentation about its use. I think this should be passed as an additional argument to `withExtensions` and `buildEnv.extensions`, optimally, but that would need much more work and this should suffice for now.

**Comment by talyz** [comment]
I've added an `enabledExtensions` attribute and documentation about its use, fixed a typo in the `nextcloud` module which broke evaluation and fixed a few other minor things. The nextcloud and php tests pass.

File: pkgs/top-level/php-packages.nix
Comment (etu) [reply pkgs/top-level/php-packages.nix:624]: Yeah, not adding anything new here. But I restored previous state at least :-)

**Comment by aanderse** [comment]
> I've added an `enabledExtensions` attribute and documentation about its use, fixed a typo in the `nextcloud` module which broke evaluation and fixed a few other minor things. The nextcloud and php tests pass.

ping @FRidh @jtojnar @etu @Ma27 @flokli does `enabledExtensions` sound good to you?

File: pkgs/development/interpreters/php/default.nix
Comment (Izorkin) [inline pkgs/development/interpreters/php/default.nix:207]: This replace don't need?
Error - `./configure: line 84378: /usr/bin/file: No such file or directory`

Comment (etu) [reply pkgs/development/interpreters/php/default.nix:207]: Do you find anything that is broken due to removing this line other than this warning?

Having this line in adds `5.7M` to the closure size, and if that only is to get rid of a warning but actually not solve any issues, then we're not pulling file in.

Because 5.7M is 6% added closure size.

Comment (Izorkin) [reply pkgs/development/interpreters/php/default.nix:207]: ok

File: pkgs/development/interpreters/php/default.nix
Comment (flokli) [inline pkgs/development/interpreters/php/default.nix:198]: @etu It seems this currently causes the phpBase  to be considered (and used) when patching shebangs in CLI tools - see `${arcanist}/bin/arcanist --help`.

Comment (flokli) [reply pkgs/development/interpreters/php/default.nix:198]: See for an example before and after: https://pastebin.com/JX1N07dv

Comment (etu) [reply pkgs/development/interpreters/php/default.nix:198]: I'm thinking that this must be due to `symlinkJoin` that was introduced in https://github.com/NixOS/nixpkgs/pull/79377.

But the problem became visible now because of the slimmer base.

We also have some weird issue with `nix-shell -p php74`, we get `php74base` in reality. And it can be confirmed by looking into the `$PATH` of the shell, we don't get the `php-with-extensions` derivation that we would expect.

Comment (flokli) [reply pkgs/development/interpreters/php/default.nix:198]: @Ma27, @aanderse , could you take a look?

Comment (talyz) [reply pkgs/development/interpreters/php/default.nix:198]: It's because we inherit the `dev` output from the original php package. I'm working on a fix right now.

Comment (talyz) [reply pkgs/development/interpreters/php/default.nix:198]: I've opened #84721 for the fix.

**Comment by chpio** [comment]
hi, i have a custom extension, that i need for my script. The extension build seems to work, but it's not available in `php.packages`. Im relatively new to nix, so pardon me.

my nix shell:

```
let
	pkgs = import <nixpkgs> {};
	phpGeos = pkgs.stdenv.mkDerivation rec {
		pname = "php-geos";
		version = "1.0.0";
		meta = {
			description = "php binding for GEOS";
		};

		src = pkgs.fetchurl {
			url = "https://git.osgeo.org/gitea/geos/php-geos/archive/${version}.tar.gz";
			sha256 = "0d90z591s092ys0d9gww2xw8vsvcdl6jbc90cgc6avq27dx4xk89";
		};

		nativeBuildInputs = [ pkgs.autoconf pkgs.automake pkgs.php74base ];
		preConfigure = "phpize";

		buildInputs = [ pkgs.geos ];

		installPhase = ''
			mkdir -p $out/lib/php/extensions
			cp modules/geos.so $out/lib/php/extensions/geos.so
		'';
	};
	php = pkgs.php74.buildEnv {
		extensions = exts: pkgs.php74.enabledExtensions ++ [phpGeos];
	};
in
	pkgs.stdenv.mkDerivation {
		name = "...";
		buildInputs = [
			php
			php.packages.composer
		];
	}

```

i have assumed that all enabled extensions would be passed to packages. When i run `php -i` my ext shows up. But when running `/nix/store/pdllmyi9bdk7hc0bxgr8x01z27wgrdi9-php-with-extensions-7.4.4/bin/php -i` (the php bin that composer is run with) the ext is not listed.

**Comment by etu** [comment]
Hi @chpio!

You're about right all the way there!

Composer get's it's own environments PHP, and currently on master it doesn't use the overriden php. But this should be totally doable when we get #85026 together.

With that PR you should be able to use `customPhp.packages.composer` for instance.

This isn't something that have been possible at all before in Nix so this will be a very exciting development :-)

If you can it would be great if you can check out that PR and report back there if it works for you.

@talyz and me will probably team up and try to get that one shaped up some day as well so we can get it in.

If you want a faster workaround in the meanwhile you could make your own derivation for composer that just use your custom php since composer is only a phar anyways.

Something like this can probably work for your case:
```nix
let
  myphp = pkgs.php74; # customize here
  mycomposer = writeScriptBin "composer" ''
    #!/bin/sh
    ${myphp} ${myphp.packages.composer}/libexec/composer/composer.phar \"$@\"
  '';
in {}
```

The only thing I *know* is missing from that composer is to inject unzip into the `$PATH` in that script.

**Comment by talyz** [comment]
Hi @chpio! Can you try #85026 now? Your use case should work fine now :)

**Pull request merged**
